crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -464,8 +481,57 @@ def save_error_state(message = "")
         Rails.logger.error(message)
       end
 
-      def upgrade_compute_nodes
-        # TODO: not implemented
+      def upgrade_all_compute_nodes
+        ["kvm", "xen"].each do |virt|
+          return false unless upgrade_compute_nodes virt
+        end
+        true
+      end
+
+      def upgrade_compute_nodes(virt)
+        save_upgrade_state("Upgrading compute nodes of #{virt} type")
+        compute_nodes = NodeObject.find("roles:nova-compute-#{virt}")
+        return true if compute_nodes.empty?
+
+        controller = NodeObject.find("roles:nova-controller").first
+        if controller.nil?
+          save_error_state("No nova controller node was found!")
+          return false
+        end
+
+        # FIXME: action in this block could be executed in parallel for all compute nodes!
+        # should we use chef-client for that?
```

Comment:
I could also use `node.ssh_cmd`, which executes the script in background. But I'm also not sure how to track progress and/or possible errors... and that is also ugly for chef execution